ql/src/java/org/apache/hadoop/hive/ql/exec/repl/bootstrap/events/filesystem/DatabaseEventsIterator.java

```diff
@@ -0,0 +1,124 @@
+package org.apache.hadoop.hive.ql.exec.repl.bootstrap.events.filesystem;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.LocatedFileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.RemoteIterator;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.ql.exec.repl.bootstrap.events.BootstrapEvent;
+import org.apache.hadoop.hive.ql.exec.repl.bootstrap.load.ReplicationState;
+import org.apache.hadoop.hive.ql.parse.EximUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.apache.hadoop.hive.ql.parse.ReplicationSemanticAnalyzer.FUNCTIONS_ROOT_DIR_NAME;
+
+class DatabaseEventsIterator implements Iterator<BootstrapEvent> {
+  private static Logger LOG = LoggerFactory.getLogger(DatabaseEventsIterator.class);
+  private RemoteIterator<LocatedFileStatus> remoteIterator;
+
+  private final Path dbLevelPath;
+  private HiveConf hiveConf;
+  ReplicationState replicationState;
+  private Path next = null, previous = null;
+  private boolean databaseEventProcessed = false;
+
+  DatabaseEventsIterator(Path dbLevelPath, HiveConf hiveConf) throws IOException {
+    this.dbLevelPath = dbLevelPath;
+    this.hiveConf = hiveConf;
+    FileSystem fileSystem = dbLevelPath.getFileSystem(hiveConf);
+    // this is only there for the use case where we are doing table only replication and not database level
+    if (!fileSystem.exists(new Path(dbLevelPath + Path.SEPARATOR + EximUtil.METADATA_NAME))) {
+      databaseEventProcessed = true;
+    }
+    remoteIterator = fileSystem.listFiles(dbLevelPath, true);
+  }
+
+  @Override
+  public boolean hasNext() {
+    try {
+      if (!databaseEventProcessed) {
+        next = dbLevelPath;
+        return true;
+      }
+
+      if (replicationState == null && next == null) {
+        while (remoteIterator.hasNext()) {
+          LocatedFileStatus next = remoteIterator.next();
+          if (next.getPath().toString().endsWith(EximUtil.METADATA_NAME)) {
+            String replacedString = next.getPath().toString().replace(dbLevelPath.toString(), "");
+            List<String> filteredNames = Arrays.stream(replacedString.split(Path.SEPARATOR))
```

Comment:
Sure, also need to check with Venkat/Thejas if this is ok as the load progress logs will have logs for tables/functions comes jumbled.